app/assets/javascripts/validation.js

```diff
@@ -10,69 +10,228 @@ $(document).ready(function () {
+  );
+  var emailElement = document.querySelector(
+    formClass + " [name='user[email]']"
+  );
+  var passwordElement = document.querySelector(
+    formClass + " [name='user[password]']"
+  );
+  var confirmPasswordElement = document.querySelector(
+    formClass + " [name='user[password_confirmation]']"
+  );
+
+  // Every time user types something, it triggers corresponding event listener
+  usernameElement.addEventListener(
+    'input',
+    validateUsername.bind(usernameElement, this)
+  );
+  emailElement.addEventListener(
+    'input',
+    validateEmail.bind(emailElement, this)
+  );
+  passwordElement.addEventListener(
+    'input',
+    validatePassword.bind(passwordElement, confirmPasswordElement, this)
+  );
+  confirmPasswordElement.addEventListener(
+    'input',
+    validateConfirmPassword.bind(confirmPasswordElement, passwordElement, this)
+  );
+}
+
+// Every time user types something in the form, it triggers this function
+// It will update UI depending on the value of <valid> parameter
+SignUpFormValidator.prototype.updateUI = function(element, valid, errorMsg) {
+  var elementName = element.getAttribute('name');
+
+  if (valid) {
+    this.validationTracker[elementName] = true;
+    styleElement(element, 'form-element-invalid', 'form-element-valid');
+    removeErrorMsg(element);
+  } else {
+    this.validationTracker[elementName] = false;
+    styleElement(element, 'form-element-valid', 'form-element-invalid');
+    renderErrorMsg(element, errorMsg);
+  }
+
+  this.isFormValid();
+};
+
+SignUpFormValidator.prototype.disableSubmitBtn = function() {
+  this.submitBtn.setAttribute('disabled', '');
+};
+
+SignUpFormValidator.prototype.enableSubmitBtn = function() {
+  this.submitBtn.removeAttribute('disabled');
+};
+
+SignUpFormValidator.prototype.isFormValid = function() {
+  // Form is valid if all elements have passsed validation successfully
+  var isValid =
+    Object.values(this.validationTracker).filter(Boolean).length === 4;
+
+  if (isValid) this.enableSubmitBtn();
+  else this.disableSubmitBtn();
+};
+
+function validateUsername(obj) {
+  var username = this.value;
+  var self = this;
+
+  if (username.length < 3) {
+    restoreOriginalStyle(this);
+  } else {
+    $.get('/api/srch/profiles?query=' + username, function(data) {
+      if (data.items) {
+        $.map(data.items, function(userData) {
+          if (userData.doc_title === username) {
+            obj.updateUI(self, false, 'Username already exists');
+          } else {
+            obj.updateUI(self, true);
+          }
+        });
+      } else {
+        obj.updateUI(self, true);
       }
-    },
-    submitHandler: function(form) {
-      form.submit();
-    }
-  });
+    });
+  }
+}
+
+function validateEmail(obj) {
+  var email = this.value;
+  var regexp = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```

Comment:
Try to use explicit variable names like emailRegex. I can't understand which regex are you talking about if I am new developer. I hope you understand. Do similarly at other places

Reply:
:+1:

Reply:
Please ensure that the regex used here and at the backend are SAME.

Reply:
I am unable to find the email regex in RoR. Maybe it's not there? Some functionality is not in this repo, but in a "database-backed something"  as @jywarren described it.

Reply:
Hi! I hope to catch up on this PR today, but need to go eat some lunch first! For now i hope this helps: the regex is generally in the Authlogic Ruby gem, but we've overridden it here: https://github.com/publiclab/plots2/blob/d251b5ff71036060d2439128496550c022bac80c/app/models/user.rb#L10

I actually am not sure why we did that. But we did. :-)

We could look back in the history to see what the justification was, but it seems out of scope for this use case. Let's just be sure people seeing that line can find their way to the JS code, and vice versa, so that if either one is ever changed, the developer knows to change the corresponding one. Thanks!

Reply:
This the username validation regex, but I need a regex that validates the email address.
It's simply not there :(